hooks/tk-multi-publish2/basic/publish_session.py

```diff
@@ -0,0 +1,712 @@
+                        "label": "Show Conflicts",
+                        "tooltip": "Show the conflicting publishes in Shotgun",
+                        "text": conflict_info
+                    }
+                }
+            )
+
+        # check to see if the next version of the work file already exists on
+        # disk. if so, warn the user and provide the ability to jump to save
+        # to that version now
+        (next_version_path, version) = self._get_next_version_info(
+            path, settings)
+        if next_version_path and os.path.exists(next_version_path):
+
+            # determine the next available version_number. just keep asking for
+            # the next one until we get one that doesn't exist.
+            while os.path.exists(next_version_path):
+                (next_version_path, version) = self._get_next_version_info(
+                    next_version_path, settings)
+
+            self.logger.error(
+                "The next version of this file already exists on disk.",
+                extra={
+                    "action_button": {
+                        "label": "Save to v%s" % (version,),
+                        "tooltip": "Save to the next available version number, "
+                                   "v%s" % (version,),
+                        "callback": lambda: _save_session(next_version_path)
+                    }
+                }
+            )
+            return False
+
+        # store the session path since this is guaranteed to run just before
+        # the publish itself
+        item.properties["work_file_path"] = path
+
+        # store the publish info in the properties
+        item.properties["publish_info"] = publish_info
+
+        self.logger.info("A Publish will be created in Shotgun and linked to:")
+        self.logger.info("  %s" % (path,))
+
+        return True
+
+    def publish(self, settings, item):
+        """
+        Executes the publish logic for the given item and settings.
+
+        :param settings: Dictionary of Settings. The keys are strings, matching
+            the keys returned in the settings property. The values are `Setting`
+            instances.
+        :param item: Item to process
+        """
+
+        publisher = self.parent
+
+        # get the work file path stored during validation
+        path = item.properties["work_file_path"]
+
+        # ensure the session is saved
+        _save_session(path)
+
+        # get all the publish info extracted during validation
+        publish_info = item.properties["publish_info"]
+        version_number = publish_info["publish_version"]
+        publish_path = publish_info["publish_path"]
+        publish_name = publish_info["publish_name"]
+        publish_type = publish_info["publish_type"]
+
+        # if the path doesn't match the publish path, copy the file to the
+        # publish path
+        if path != publish_path and not os.path.exists(publish_path):
```

Comment:
This is copying the workfile to the publish file path. It's not explicitly using templates here, but if they are different paths, then it can be inferred. Any concerns here?

Reply:
Only issue I see is if the publish does already exist this doesn't warn or anything... or would that fail validation?

Reply:
it's not uncommon that you have a workfile names `{name}.{version].ma` and a published file named `{shot}_{name}.{version}.ma` - when a file is used for downstream consumption, sometimes studios (sensibly!) want to attach additional state to the file name (e.g. shot name). So i don't think a copy will work - need to extract the fields from the work path and then use those fields + the context to assemble the publish template.

Reply:
That's all done in the method call above that figures out what the publish path should be. It's either template based (if they're configured and defined) or it uses the basic workflow as a fallback.